### PR TITLE
Add a new login component UI

### PR DESCRIPTION
**Pull request by hatched**
This branch removes the old login components and adds a new React based one. It has no UI beyond two inputs and a submit button. A UI update will land once designs are supplied.

### CODE REVIEW

**Comment by hatched** [comment]
#### To QA

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3738/
Test FAILed.

File: jujugui/static/gui/src/app/store/env/base.js [hunk 1]
```diff
@@ -364,8 +364,7 @@ YUI.add('juju-env-base', function(Y) {
              * @return {Boolean} Whether or not user and password are set.
              */
             get: function() {
-              return Y.Lang.isValue(this.user) &&
-                  Y.Lang.isValue(this.password);
+              return !!this.user && !!this.password;
```
Comment (hatched) [inline jujugui/static/gui/src/app/store/env/base.js]: This would treat boolean false values as valid user and password values.

File: jujugui/static/gui/src/app/utils/component-test-utils.js [hunk 2]
```diff
@@ -99,10 +99,11 @@ var jsTestUtils = {
     @method compare
     @param {Object} a The source object.
     @param {Object} b The expected object.
+    @param {Boolean} showFn Whether you want to show the functions.
   */
-  compare: function compare(a, b) {
-    var stringA = jsTestUtils.superStringify(a);
-    var stringB = jsTestUtils.superStringify(b);
+  compare: function compare(a, b, showFn) {
```
Comment (hatched) [inline jujugui/static/gui/src/app/utils/component-test-utils.js]: driveby to also be able to compare the functions

File: jujugui/static/gui/src/app/views/state.js [hunk 3]
```diff
@@ -108,7 +108,9 @@ YUI.add('juju-app-state', function(Y) {
         if (this.hasChanged(section, 'component')) {
           this._emptySection(section);
         }
-        this._dispatchSection(section, state[section]);
+        if (state[section]) {
```
Comment (hatched) [inline jujugui/static/gui/src/app/views/state.js]: driveby - state was dispatching all sections even if none had changed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3739/
Test FAILed.

File: jujugui/static/gui/src/app/views/utils.js [hunk 4]
```diff
@@ -1513,15 +1513,15 @@ YUI.add('juju-view-utils', function(Y) {
         } else {
           charm = charm[0];
           var config = {},
-              options = charm.get('options');
+              options = charm.get ? charm.get('options') : charm.options;
```
Comment (hatched) [inline jujugui/static/gui/src/app/views/utils.js]: I wasn't really sure if this can ever be a Y.Model anymore with the new jujulib changes so I just made it accept both just to be safe.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3740/
Test FAILed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3741/
Test PASSed.

File: jujugui/static/gui/src/app/components/login/login.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+YUI.add('login-component', function() {
+
+  juju.components.Login = React.createClass({
+
+    propTypes: {
+      setCredentials: React.PropTypes.func.isRequired,
+      login: React.PropTypes.func.isRequired
+    },
+
+    _handleSubmit: function(e) {
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/login/login.js]: Doc comments.

File: jujugui/static/gui/src/app/views/state.js [hunk 6]
```diff
@@ -464,7 +470,12 @@ YUI.add('juju-app-state', function(Y) {
           state.sectionA = this._addToSection({
             component: 'services'
           });
-        } else if (part.indexOf('profile') === 0) {
+        } else if (part.indexOf('login') === 0) {
+          state.app = this._addToSection({
+            component: 'login'
+          });
+        }
+        else if (part.indexOf('profile') === 0) {
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/views/state.js]: `} else if ...`

**Comment by huwshimi** [comment]
QA OK :+1: Great stuff @hatched.

File: jujugui/static/gui/src/app/views/utils.js (hunk 4 above)
Comment (jaycee) [inline jujugui/static/gui/src/app/views/utils.js]: I believe any data out of jujulib that the gui gets is converted to a model; if it isn't that should probably be marked as a bug.

**Comment by jaycee** [comment]
QA:

Sandbox mode, login worked as described in QA steps.

In the charm, it allows logging in and getting to the canvas; there continue to be errors, but that appears to be related to the charm setup as it exists w/ or w/o this version of the gui.

QA OK.

File: jujugui/static/gui/src/app/components/login/login.js (hunk 5 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/login/login.js]: done

File: jujugui/static/gui/src/app/views/state.js (hunk 6 above)
Comment (hatched) [inline jujugui/static/gui/src/app/views/state.js]: done

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3746/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3747/
Test FAILed.

**Comment by hatched** [comment]
Thanks for the reviews! :shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3748/
Test PASSed.
